src/token/api/src/state/balances.rs

```diff
@@ -0,0 +1,229 @@
+pub struct LocalBalances(HashMap<AccountInternal, Tokens128>);
+
+impl LocalBalances {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+}
+
+impl FromIterator<(AccountInternal, Tokens128)> for LocalBalances {
+    fn from_iter<T: IntoIterator<Item = (AccountInternal, Tokens128)>>(iter: T) -> Self {
+        Self(HashMap::from_iter(iter))
+    }
+}
+
+impl Balances for LocalBalances {
+    fn insert(&mut self, account: AccountInternal, token: Tokens128) {
+        self.0.insert(account, token);
+    }
+
+    fn get(&self, account: &AccountInternal) -> Option<Tokens128> {
+        self.0.get(account).copied()
+    }
+
+    fn list_balances(&self, start: usize, limit: usize) -> Vec<(AccountInternal, Tokens128)> {
+        let mut holders = self
+            .0
+            .iter()
+            .skip(start)
+            .take(limit)
+            .map(|(account, tokens)| (*account, *tokens))
+            .collect::<Vec<_>>();
+        holders.sort_by(|a, b| b.1.cmp(&a.1));
+        holders
+    }
+
+    fn remove(&mut self, account: &AccountInternal) -> Option<Tokens128> {
+        self.0.remove(account)
+    }
+
+    fn total_supply(&self) -> Tokens128 {
+        self.0
+            .iter()
+            .fold(Tokens128::ZERO, |a, b| (a + b.1).unwrap_or(Tokens128::ZERO))
+    }
+}
+
+const BALANCES_MEMORY_ID: MemoryId = MemoryId::new(1);
+
+struct Key {
+    pub principal: Principal,
+    pub subaccount: Subaccount,
+}
+
+impl From<AccountInternal> for Key {
+    fn from(account: AccountInternal) -> Self {
+        Self {
+            principal: account.owner,
+            subaccount: account.subaccount,
+        }
+    }
+}
+
+impl From<Key> for AccountInternal {
+    fn from(key: Key) -> Self {
+        Self {
+            owner: key.principal,
+            subaccount: key.subaccount,
+        }
+    }
+}
+
+const SUBACCOUNT_LEN: usize = 32;
+const SUBACCOUNT_OFFSET: usize = 32;
+
+const KEY_BYTES_LEN: usize = 64;
+const VALUE_BYTES_LEN: usize = 16;
+
+impl Storable for Key {
+    /// Memory layout:
+    /// | principal len (1 byte) | principal data (31 byte) | subaccount (32 bytes) |
+    /// Principal data maximum len is 29;
+    fn to_bytes(&self) -> Cow<[u8]> {
+        let mut buffer = vec![0u8; KEY_BYTES_LEN];
+        let principal_bytes = self.principal.as_slice();
+
+        buffer[0] = principal_bytes.len() as u8;
+        buffer[1..principal_bytes.len() + 1].copy_from_slice(principal_bytes);
+        buffer[SUBACCOUNT_OFFSET..].copy_from_slice(&self.subaccount);
+
+        Cow::Owned(buffer)
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        let principal_len = bytes[0] as usize;
+        let account = Principal::from_slice(&bytes[1..principal_len + 1]);
+        let mut subaccount = [0u8; SUBACCOUNT_LEN];
+        subaccount.copy_from_slice(&bytes[SUBACCOUNT_OFFSET..]);
+        Self {
+            principal: account,
+            subaccount,
+        }
+    }
+}
```

Comment:
Let's add a couple unit tests for this. This is too scary to leave without dedicated unit tests.

Reply:
Done